### PR TITLE
feat(tabby): require git_url for snippet search

### PR DESCRIPTION
**Pull request by boxbeam**
Fixes TAB-574

### CODE REVIEW

File: crates/tabby/src/services/completion/completion_prompt.rs [hunk 1]
```diff
@@ -99,12 +104,17 @@ fn build_prefix(language: &str, prefix: &str, snippets: &[Snippet]) -> String {
...
+    git_url: &str,
+    language: &str,
+    text: &str,
+) -> Vec<Snippet> {
     let mut ret = Vec::new();
     let mut tokens = tokenize_text(text);
 
     let serp = match code
-        .search_in_language(language, &tokens, MAX_SNIPPETS_TO_FETCH, 0)
+        .search_in_language(git_url, language, &tokens, MAX_SNIPPETS_TO_FETCH, 0)
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion/completion_prompt.rs:117]: implement some heurstic based normalization to git url, for example:
1. git url being managed in repositories contains `https://github.com/TabbyML/tabby`
2. user's local git clone is from `https://github.com/TabbyML/tabby.git`

I suggest implement some herustic match like tokenization into [github.com, TabbyML, tabby], [github.com, TabbyML, tabby.git], and requires certain matching.

Comment (boxbeam) [reply crates/tabby/src/services/completion/completion_prompt.rs:117]: Note: Use non-exact string match rather than sanitizing input

File: crates/tabby-common/src/index.rs [hunk 2]
```diff
@@ -101,6 +98,38 @@ impl CodeSearchSchema {
                 .map(|x| Term::from_field_text(self.field_body, x)),
         ))
     }
+
+    pub fn git_url_query(&self, mut git_url: &str) -> Vec<Box<TermQuery>> {
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index.rs:102]: this won't work: git_url field is indexed as a whole token, thus can only matched against exact url.

I suggest to use a normalization step by matching git_url against all possible git_urls (from db), and used the matched git_url in tantivy query.

File: ee/tabby-webserver/src/handler.rs [hunk 3]
```diff
@@ -1,51 +1,107 @@
...
+        Ok(repos)
+    }
+}
+
+fn new_repository_access(service: Arc<dyn RepositoryService>) -> impl RepositoryAccess {
+    ServiceRepositoryAccess {
+        service,
+        url_cache: Mutex::new(TimedCache::with_lifespan(10 * 60)),
+    }
+}
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/handler.rs:73]: merge into function `repository_access`?

File: ee/tabby-webserver/src/handler.rs [hunk 4]
```diff
@@ -1,51 +1,107 @@
...
+    repositories,
+    schema::{
+        auth::AuthenticationService, create_schema, repository::RepositoryService, Schema,
+        ServiceLocator,
+    },
     service::{create_service_locator, event_logger::create_event_logger},
     ui,
 };
 
+struct ServiceRepositoryAccess {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/handler.rs:39]:
```suggestion
struct RepositoryAccessImpl {
```

File: crates/tabby/src/services/completion.rs [hunk 5]
```diff
@@ -294,11 +301,15 @@ impl CompletionService {
 
         let (prompt, segments, snippets) = if let Some(prompt) = request.raw_prompt() {
             (prompt, None, vec![])
-        } else if let Some(segments) = request.segments.clone() {
+        } else if let Some((segments, git_url)) =
+            request.segments.as_ref().zip(request.git_url.as_ref())
+        {
+            debug!("PREFIX: {}, SUFFIX: {:?}", segments.prefix, segments.suffix);
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:307]:
```suggestion
```

File: crates/tabby/src/services/completion.rs [hunk 6]
```diff
@@ -309,9 +320,10 @@ impl CompletionService {
         } else {
             return Err(CompletionError::EmptyPrompt);
         };
+        debug!("PROMPT: {}", prompt);
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:323]:
```suggestion
```

File: crates/tabby/src/services/completion.rs [hunk 7]
```diff
@@ -309,9 +320,10 @@ impl CompletionService {
...
 
         let text = self.engine.generate(&prompt, options).await;
-        let segments = segments.map(|s| s.into());
+        let segments = segments.cloned().map(|s| s.into());
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:326]: is cloned here necessary?

File: crates/tabby/src/services/completion.rs [hunk 8]
```diff
@@ -52,6 +53,9 @@ pub struct CompletionRequest {
 
     /// The seed used for randomly selecting tokens
     seed: Option<u64>,
+
+    /// The URL of the git repository being worked on
+    git_url: Option<String>,
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:58]:
```suggestion
```

Use `segments.git_url`

File: crates/tabby/src/serve.rs [hunk 9]
```diff
@@ -130,19 +130,26 @@ pub async fn main(config: &Config, args: &ServeArgs) {
     };
 
     let logger: Arc<dyn EventLogger>;
+    let repo_access: Arc<dyn RepositoryAccess>;
```
Comment (wsxiaoys) [inline crates/tabby/src/serve.rs:133]:
```suggestion
    let repository_access: Arc<dyn RepositoryAccess>;
```

File: crates/tabby/src/services/code.rs [hunk 10]
```diff
@@ -22,10 +23,11 @@ struct CodeSearchImpl {
     query_parser: QueryParser,
 
     schema: CodeSearchSchema,
+    repo_access: Arc<dyn RepositoryAccess>,
 }
 
 impl CodeSearchImpl {
-    fn load() -> Result<Self> {
+    fn load(repo_access: Arc<dyn RepositoryAccess>) -> Result<Self> {
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:30]:
```suggestion
    fn load(repository_access: Arc<dyn RepositoryAccess>) -> Result<Self> {
```

File: crates/tabby/src/services/code.rs [hunk 11]
```diff
@@ -40,15 +42,16 @@ impl CodeSearchImpl {
...
         Ok(Self {
+            repo_access,
             reader,
             query_parser,
             schema: code_schema,
         })
     }
 
-    async fn load_async() -> CodeSearchImpl {
+    async fn load_async(repo_access: Arc<dyn RepositoryAccess>) -> CodeSearchImpl {
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:52]: ditto

File: crates/tabby/src/services/code.rs [hunk 12]
```diff
@@ -22,10 +23,11 @@ struct CodeSearchImpl {
     query_parser: QueryParser,
 
     schema: CodeSearchSchema,
+    repo_access: Arc<dyn RepositoryAccess>,
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:26]:
```suggestion
    repository_access: Arc<dyn RepositoryAccess>,
```

File: crates/tabby/src/services/completion.rs [hunk 13]
```diff
@@ -247,6 +251,7 @@ impl CompletionService {
 
     async fn build_snippets(
         &self,
+        git_url: &str,
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:254]:
```suggestion
```

use `segments.git_url`

File: crates/tabby-common/src/index.rs [hunk 14]
```diff
@@ -101,6 +101,34 @@ impl CodeSearchSchema {
...
+                    // haystack = "https://abc@github.com/boxbeam/untwine.git" needle = "https://github.com/boxbeam/untwine" => Match, score 842
+                    nucleo.fuzzy_match(
+                        Utf32String::from(git_url).slice(..),
+                        Utf32String::from(&*repo.git_url).slice(..),
+                    )?,
+                ))
+            })
+            .max_by_key(|(_, score)| *score)
+            .map(|(entry, _score)| entry)
+            .unwrap_or_else(|| git_url.to_string());
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index.rs:125]: move fuzzy logic to `service/code.rs`

File: crates/tabby-common/src/index.rs [hunk 15]
```diff
@@ -101,6 +101,34 @@ impl CodeSearchSchema {
...
+            .filter_map(|repo| {
+                Some((
+                    repo.git_url.clone(),
+                    // Matching using the input URL as the haystack instead of the needle yielded better scoring
+                    // Example:
+                    // haystack = "https://github.com/boxbeam/untwine" needle = "https://abc@github.com/boxbeam/untwine.git" => No match
+                    // haystack = "https://abc@github.com/boxbeam/untwine.git" needle = "https://github.com/boxbeam/untwine" => Match, score 842
+                    nucleo.fuzzy_match(
+                        Utf32String::from(git_url).slice(..),
+                        Utf32String::from(&*repo.git_url).slice(..),
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index.rs:119]: strip credential part of git_url before matching.

File: crates/tabby-common/src/index.rs [hunk 16]
```diff
@@ -101,6 +101,34 @@ impl CodeSearchSchema {
...
+                    )?,
+                ))
+            })
+            .max_by_key(|(_, score)| *score)
+            .map(|(entry, _score)| entry)
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index.rs:124]: apply threshold on score (and add some unit test to verify the threshold)

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/handler.rs (hunk 4 above)
Comment (boxbeam) [reply ee/tabby-webserver/src/handler.rs:39]: Didn't want to call it this because we have other implementations of `RepositoryAccess`, I think this is more clear since it is the implementation of `RepositoryAccess` which is based on the service as opposed to config.

File: crates/tabby/src/services/completion.rs (hunk 7 above)
Comment (boxbeam) [reply crates/tabby/src/services/completion.rs:326]: Yes

File: ee/tabby-webserver/src/handler.rs (hunk 4 above)
Comment (wsxiaoys) [reply ee/tabby-webserver/src/handler.rs:39]: Service is too generic - since this srruct is private to the file, `Impl` should be a sufficient good default.

File: crates/tabby/src/services/code.rs [hunk 17]
```diff
@@ -177,16 +228,57 @@ impl CodeSearch for CodeSearchService {
...
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_closest_match() {
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:251]: add negative test cases? (closest match is None)

File: crates/tabby/src/services/code.rs [hunk 18]
```diff
@@ -113,21 +118,67 @@ impl CodeSearch for CodeSearchImpl {
...
+            (Occur::Must, git_url_query),
         ]);
         self.search_with_query(&query, limit, offset).await
     }
 }
 
+fn closest_match<'a>(
+    search_term: &'a str,
+    search_input: impl IntoIterator<Item = &'a str>,
+) -> String {
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:160]:
```suggestion
) -> Option<String> {
```

File: crates/tabby/src/services/code.rs [hunk 19]
```diff
@@ -113,21 +118,67 @@ impl CodeSearch for CodeSearchImpl {
...
+                // haystack = "https://abc@github.com/boxbeam/untwine.git" needle = "https://github.com/boxbeam/untwine" => Match, score 842
+                nucleo.fuzzy_match(
+                    Utf32String::from(search_term).slice(..),
+                    Utf32String::from(entry).slice(..),
+                )?,
+            ))
+        })
+        .max_by_key(|(_, score)| *score)
+        .map(|(entry, _score)| entry)
+        .unwrap_or_else(|| search_term.to_string())
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:179]: return none if there is no match

File: crates/tabby/src/services/code.rs [hunk 20]
```diff
@@ -113,21 +118,67 @@ impl CodeSearch for CodeSearchImpl {
...
+            .await
+            .unwrap_or_default();
+
+        let git_url = Url::parse(git_url)
+            .map(|mut url| {
+                let _ = url.set_password(None);
+                let _ = url.set_username("");
+                url.to_string()
+            })
+            .unwrap_or_else(|_| git_url.to_string());
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:142]: 1. move stripping logic as part of `closest_match`.
2. git_urls from `RepositoryAccess` also requires stripping.

File: crates/tabby/src/services/code.rs [hunk 21]
```diff
@@ -113,21 +118,67 @@ impl CodeSearch for CodeSearchImpl {
...
+
+        let git_url = closest_match(&git_url, repos.iter().map(|repo| &*repo.git_url))
+            .unwrap_or_else(|| git_url.to_string());
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:145]: When it's none, there's no need to do the search, you can early return with empty result.

File: crates/tabby/src/services/code.rs [hunk 22]
```diff
@@ -177,16 +228,57 @@ impl CodeSearch for CodeSearchService {
...
+            ),
+            Some("https://github.com/examp/anoth-repo".into())
+        );
+
+        assert_eq!(
+            closest_match(
+                "https://github.com/TabbyML/tabby",
+                ["https://github.com/TabbyML/registry-tabby"]
+            ),
+            None
```
Comment (wsxiaoys) [inline crates/tabby/src/services/code.rs:281]: please add more test cases expecting a None result, even just for sanity check.

File: crates/tabby-common/src/index.rs [hunk 23]
```diff
@@ -101,6 +98,13 @@ impl CodeSearchSchema {
                 .map(|x| Term::from_field_text(self.field_body, x)),
         ))
     }
+
+    pub fn git_url_query(&self, git_url: &str) -> Box<TermQuery> {
```
Comment (wsxiaoys) [inline crates/tabby-common/src/index.rs:102]: add unit test for coverage

**Review by wsxiaoys — APPROVED** [review]

File: crates/tabby-common/src/index.rs (hunk 23 above)
Comment (boxbeam) [reply crates/tabby-common/src/index.rs:102]: Not sure what there is to test with this function?
